runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/state/StateRequestHandlers.java

```diff
@@ -214,17 +258,22 @@ public static StateRequestHandler forMultimapSideInputHandlerFactory(
             TypeCase.MULTIMAP_SIDE_INPUT);
 
         StateKey.MultimapSideInput stateKey = request.getStateKey().getMultimapSideInput();
-        MultimapSideInputSpec<?, ?, ?> sideInputReferenceSpec =
+        MultimapSideInputSpec<Object, Object, BoundedWindow> referenceSpec =
             processBundleDescriptor.getMultimapSideInputSpecs()
                 .get(stateKey.getPtransformId())
                 .get(stateKey.getSideInputId());
-        MultimapSideInputHandler<?, ?, ?> handler = cache.computeIfAbsent(
-            sideInputReferenceSpec,
+        MultimapSideInputHandler<Object, Object, BoundedWindow> handler = cache.computeIfAbsent(
+            referenceSpec,
             this::createHandler);
 
+        // TODO: Migrate to using the ByteStringCoder everywhere for user types and values
+        // to reduce the amount of byte array copying.
+        Object key = referenceSpec.keyCoder().decode(stateKey.getKey().newInput());
```

Comment:
Does this mean only key types with built-in coders are supported? Why do we need to decode the key here?

Reply:
No, all types are supported. We are hiding the fact that we are using the ByteStringCoder internally (hardcoded in ProcessBundleDescriptors#forBagUserState) as an optimization.

This allows us to pass an Object and a Coder without the Runner needing to care how to turn that into bytes giving us the freedom to change which types we pass around without the Runner being the wiser. The TODO is about changing the interface to be tailored towards using ByteString directly on the interface and would be a minor performance improvement for some runners.